### PR TITLE
Proof of Attendance submission

### PR DESCRIPTION
**Pull request by spicycheeto**
Update to my submision. Now fully integrated with XUMM app and pinata ipfs. Requires api keys in xummApi.js. Will be updating readme and tests can be found in tests folder.

### CODE REVIEW

**Comment by jonathanlei** [comment]
Thanks for submitting, will take a look and leave comments/revisions by EOW

File: frontend2.0/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Getting Started with Create React App
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+
+## Available Scripts
+
+In the project directory, you can run:
```
Comment (jonathanlei) [inline frontend2.0/README.md]: could you add `npm install`?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "mern-boilerplate",
```
Comment (jonathanlei) [inline package.json]: should this file be in the backend folder? is there a README explaining the technical design and instructions to run backend + frontend?

File: backend/routes/XummApi.js [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+const xummCall = async (payload) => {
+
+    const data =  await sdk.payload.create(payload)
+    return data
+}
+
+
+
+router.route('/mintTickets').post( (req,res) => {
```
Comment (jonathanlei) [inline backend/routes/XummApi.js:50]: Each route would need documentation on what params/ metadata is needed, because if this submission is to be accepted as the official version, full API reference need to be provided for event organizers.

File: backend/routes/XummApi.js [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+    xummCall(payload).then( (xummInfo) => {
+        
+
+        let response = JSON.stringify(xummInfo); 
+       
+        res.send(response);
+    
+    })
+
```
Comment (jonathanlei) [inline backend/routes/XummApi.js]: why are there so many empty line?

File: backend/routes/readme.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Full readme and front end tests can be found here - still under construction:
+
+https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure
```
Comment (jonathanlei) [inline backend/routes/readme.md]: could you change this? since we are moving to xrpl bounties repo

File: backend/routes/readme.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+Full readme and front end tests can be found here - still under construction:
+
+https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure
+
+notes: 
+
+-memos are not attaching but single minting is coming back successful.
```
Comment (jonathanlei) [inline backend/routes/readme.md]: are those still the case?

File: frontend2.0/README.md (hunk 1 above)
Comment (jonathanlei) [reply frontend2.0/README.md]: seems like this README is copied from create react app, could you edit it to explain more of the frontend design?

**Review by jonathanlei — CHANGES_REQUESTED** [review]

File: backend/routes/XummApi.js (hunk 4 above)
Comment (spicycheeto) [reply backend/routes/XummApi.js]: Fixed. Left overs from debugging.

File: backend/routes/readme.md (hunk 5 above)
Comment (spicycheeto) [reply backend/routes/readme.md]: Updated all readme files.

File: backend/routes/XummApi.js (hunk 3 above)
Comment (spicycheeto) [reply backend/routes/XummApi.js:50]: Comments have been added with proper param requirements as well as return results in xummapi.js. Examples of post / get requests with expected metadata body object can be found in /tests/app.test.js. I can add the contents of the comments in xummapi.js to the readme if needed.

File: package.json (hunk 2 above)
Comment (spicycheeto) [reply package.json]: Package.json files have been updated. Now reads "Proof-of-attendance-backend"

File: frontend2.0/README.md (hunk 1 above)
Comment (spicycheeto) [reply frontend2.0/README.md]: Readme files have been updated.

File: backend/routes/readme.md (hunk 6 above)
Comment (spicycheeto) [reply backend/routes/readme.md]: Yes, I believe this is an issue with the xumm walllet not displaying it. The memo is attached properly but doesnt come up in searches.

**Comment by spicycheeto** [comment]
> Thanks for submitting, will take a look and leave comments/revisions by EOW

Sorry, still getting used to navigating github. I believe I resolved all your comments. Awaiting further instruction.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,82 @@
 # Proof-of-Attendance
 Community Event Management Tool: Build a library to provide infrastructure that allows event organizers to mint and distribute Attendance NFTs on the XRP Ledger.
+
+XProof-of-Attendance-Infrastructure is a MERN stack application that contains functionality to allow event organizers to mint and distribute Attendance NFTs on the XRP Ledger. The core parts of this application consist of a React frontend, Express backend, Rest API (nftRoutes.js and xummapi.js), and XrplNFTHelper.js.
```
Comment (JST5000) [inline README.md:4]: nit: It might be good to use a consistent capitalization scheme for your files.

File: tests/App.test.js [hunk 8]
```diff
@@ -0,0 +1,308 @@
+
+
+
+import axios from 'axios'; 
+import assert from 'assert';
+
+
+//Mint NFT test
+function test1() {
```
Comment (JST5000) [inline tests/App.test.js:9]: Instead of just commenting the more detailed test information, I'd recommend just using the comment as the name. So in this case `test1` -> `mintNftTest`

File: README.md [hunk 9]
```diff
@@ -1,2 +1,82 @@
...
+
+In order to run a test of all backend api routes please use the following command:
+### `npm test`
+
+This will run 'node /tests/App.test.js'
+
+# Test reference (examples of proper post/get body object for each route):
+
+test1()
+Mint NFT test
```
Comment (JST5000) [inline README.md:45]: You can add relative links here so readers can directly click to go to the test you're talking about :) - https://stackoverflow.com/questions/7653483/github-relative-link-in-markdown-file

File: backend/config/db.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import mongoose from 'mongoose'
+import user from '../models/usersModel.js'
+
+const connectDB = async () => {
+    try {
+        //database Name
```
Comment (JST5000) [inline backend/config/db.js:6]:
```suggestion
```
nit: Name of variable says it all

File: backend/routes/XrplNFTHelper.js [hunk 11]
```diff
@@ -0,0 +1,462 @@
+import xrpl from 'xrpl';
+import axios from 'axios';
+import fs  from 'fs'; //used for storage of token metadata in devmode
+
+import XummSdk from 'xumm-sdk';
+
+
+
+const sdk = new XummSdk.XummSdk('d16951f8-6528-4c3f-88e9-37333e383421','e6d8a068-a55b-4e5a-b540-016847d7e376')
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js]: Careful, I think this is a secret listed in a public repo - you'll want to use environment variables here instead. (Since GitHub stores history long-term, you'll want to get a new secret issued if you can, or go back and delete commits with this information in it)

You'll want to use a `.env` file like this: https://github.com/XRPL-Labs/XUMM-SDK/blob/master/.env.sample according to the docs on the npm page here: https://www.npmjs.com/package/xumm-sdk

Comment (JST5000) [reply backend/routes/XrplNFTHelper.js]: If you do use a `.env` file, make sure to include the steps for how to set it up clearly in the README (otherwise it can be really confusing to figure out what's broken in the configuration & how to fix it)

File: backend/routes/XrplNFTHelper.js [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+		NFTokenTaxon: this.transactionDetails.NFTokenTaxon,
+    }
+    
+    //submit minting transaction
+    const tx = await client.submitAndWait(transactionData,{wallet})
+
+  const result = await client.request({
+    method: "account_nfts",
+    account: this.transactionDetails.Account
+    })
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js:62]: This will only give up to 200 responses - if there are more than that issued, you'll have to page through the results or you may not get the latest nft. Also, I'm not sure that there's a guaranteed order for how NFTs are returned (at least I didn't see one listed on the docs page for it here: https://xrpl.org/account_nfts.html#account_nfts)

File: backend/routes/XrplNFTHelper.js [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+    for(let i = 0; i <= nftokenCount; i++)
+    {
+          nfts = await client.request({
+              method: "account_nfts",
+              account: wallet.classicAddress,
+              limit: 400,
+              marker: nfts.result.marker
+          })
+        
+    }
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js:200]: This will likely send more requests than you seem to want. I think you may have meant to do something like this to track all the nfts you get, and ensure you request all nfts for the account, even if there are more than 400.
```suggestion
    const limit = 400
    const max_requests = (nftokenCount / limit) * 10 // * 10 since servers do not have to honor the limit you ask for.
    for(let i = 0; i <= max_requests && nfts.length < nftokenCount; i++)
    {
          nfts += await client.request({
              method: "account_nfts",
              account: wallet.classicAddress,
              limit: 400,
              marker: nfts.result.marker
          })
    }
```

File: backend/routes/XrplNFTHelper.js [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+        const client = new xrpl.Client(this.clientDetails)
+        await client.connect()
+
+        console.log("Connected to Sandbox..getting all NFT's.****")
+  
+      
+        let nfts = await client.request({
+          method: "account_nfts",
+          account: this.transactionDetails.Account
+          })
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js:236]: same pagination potential issue

File: backend/routes/XrplNFTHelper.js [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+
+
+          const transactionData = {
+            TransactionType: this.transactionDetails.TransactionType,
+            Account: this.transactionDetails.Account,
+            NFTokenID: nfts.result.account_nfts[index].NFTokenID
+        }
+
+
+          const tx = await client.submitAndWait(transactionData,{wallet})
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js:336]: What should happen if there's an error in burning one of the transactions? (Should it keep going, or should it stop?)

File: backend/routes/XrplNFTHelper.js [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+
+    const client = new xrpl.Client(this.clientDetails) 
+
+    await client.connect()
+
+    let response = await client.request({
+      "command": "account_objects",
+      "account": this.transactionDetails.Account,
+      "type": "ticket"
+    })
```
Comment (JST5000) [inline backend/routes/XrplNFTHelper.js:380]: This will run into the pagination problem as well - might be worth making a helper which can do the pagination part for you.

File: backend/routes/XummApi.js [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+import {create} from 'ipfs-http-client';
+import XrplNFTHelper from './XrplNFTHelper.js';
+import XummSdk from 'xumm-sdk';
+
+
+const xummApi = ''
+const xummSecret = ''
+
+const pinataApiKey = ''
+const pinataSecretApiKey = ''
```
Comment (JST5000) [inline backend/routes/XummApi.js:14]: These should probably move to a `.env` file, or read from environment variables to ensure people don't accidentally leave their passwords in a plaintext file.

File: backend/routes/XummApi.js (hunk 3 above)
Comment (JST5000) [reply backend/routes/XummApi.js:50]: Adding those to the readme would be helpful I think. It can be hard to parse through a reference doc if you can't use ctrl + F over the contents.

File: backend/routes/XummApi.js [hunk 18]
```diff
@@ -0,0 +1,424 @@
...
+* @params: Post with a body object containing metadata.TicketCount. Integer value greater than 1. 
+* Example: req.body.metadata.TicketCount -> {metadata: {TicketCount: 5 }  }
+*
+* @Returns: XUMM data object. response.data.refs.qr_png contains qr code information for user to scan with app. 
+*/
+router.route('/createTickets').post((req,res) => {
+ 
+  let payload = {
+    TransactionType: "TicketCreate",
+    TicketCount: req.body.metadata.TicketCount
```
Comment (JST5000) [inline backend/routes/XummApi.js:73]: You should probably add a check here limiting the number of tickets to the max of 250 tickets per account, otherwise giving a nice error. https://xrpl.org/tickets.html#tickets

**Comment by JST5000** [comment]
I'm closing this PR as we've accepted another submission for Milestones 1-3 for the Proof of Attendance bounty.

**Comment by jonathanlei** [comment]
@spicycheeto thank you for all your hard work that you've put in so far, would love to see you contribute to different bounties in the future!
